### PR TITLE
Subscript default arguments

### PR DESCRIPTION
**Pull request by slavapestov**
https://bugs.swift.org/browse/SR-6118

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
I'm pretty sure this needs a proposal, especially since it has potentially weird implications for exclusivity.

**Comment by slavapestov** [comment]
@jrose-apple It was part of the generic subscripts proposal but never implemented; however that predates exclusivity, IIRC. What are the weird implications for exclusivity? When emitting a subscript, SILGen evaluates the entire argument list (including default args) once, then passes it to the getter and setter as needed. I didn't actually have to change anything in subscript argument emission because John redid it recently to use the same mechanism is call argument emission.

**Comment by jrose-apple** [comment]
Oh, sorry, I forgot that subscripts already can't have `inout` indices. Withdrawn. And the citation of the generic subscripts proposal does make this okay.

Can you add tests for
- inlinability checking
- parseable interface printing

to make sure this won't regress those features?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: test/decl/func/default-values-swift4.swift [hunk 1]
```diff
@@ -111,3 +111,29 @@ public func evilCode(
...
+    // expected-error@-1 2{{global function 'internalFunction()' is internal and cannot be referenced from a default argument value}}
+
+    fileprivateFunction()
+    // expected-error@-1 2{{global function 'fileprivateFunction()' is fileprivate and cannot be referenced from a default argument value}}
+
+    privateFunction()
+    // expected-error@-1 2{{global function 'privateFunction()' is private and cannot be referenced from a default argument value}}
+
+    return 0
+  }()) -> Int {
```
Comment (jrose-apple) [inline test/decl/func/default-values-swift4.swift]: Am I paranoid for wondering if there might be a difference when not using a closure?

File: test/ParseableInterface/default-args.swift [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+// CHECK: func hasClosureDefaultArg(_ x: () -> Void = {
+// CHECK-NEXT: })
+public func hasClosureDefaultArg(_ x: () -> Void = {
+}) {
+}
+
+// CHECK: func hasSimpleDefaultArgs(_ x: Int = 0, b: Int = 1)
+public func hasSimpleDefaultArgs(_ x: Int = 0, b: Int = 1) {
+}
```
Comment (jrose-apple) [inline test/ParseableInterface/default-args.swift]: Nitpick: Can you add a `#file` case here too, since that's serialized differently?

**Review by jrose-apple — COMMENTED** [review]
Thank you!

File: test/ParseableInterface/default-args.swift (hunk 2 above)
Comment (slavapestov) [reply test/ParseableInterface/default-args.swift]: Done

File: test/decl/func/default-values-swift4.swift (hunk 1 above)
Comment (slavapestov) [reply test/decl/func/default-values-swift4.swift]: Done

Comment (jrose-apple) [reply test/decl/func/default-values-swift4.swift]: ! That was a big improvement, but what I actually meant was "is it possible usable-from-inline checking works in the closure case but not in the `arg: Int = privateFunction()` case".

Comment (slavapestov) [reply test/decl/func/default-values-swift4.swift]: Look at the last test case in this file :)

Comment (jrose-apple) [reply test/decl/func/default-values-swift4.swift]: Whoops, paid too much attention to the rest of it!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Pull request merged**